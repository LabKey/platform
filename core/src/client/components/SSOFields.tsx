import React, { PureComponent } from 'react';
import { FileAttachmentForm } from '@labkey/components';
import { ActionURL } from '@labkey/api';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faImage, faTimesCircle } from '@fortawesome/free-solid-svg-icons';

class NoImageSelectedDisplay extends PureComponent {
    render () {
        return(
            <div className="sso-fields__null-image">
                <FontAwesomeIcon icon={faImage} color="#DCDCDC" size="6x" />
                <div className="sso-fields__null-image__text">None selected</div>
            </div>
        );
    }
}

interface ImgFileAttachForm_Props {
    text: string;
    imageUrl: string;
    onFileChange: Function;
    handleDeleteLogo: Function;
    fileTitle: string;
    canEdit: boolean;
    index: number;
}

interface ImgFileAttachForm_State {
    imageUrl: string;
}

export class ImageAndFileAttachmentForm extends PureComponent<ImgFileAttachForm_Props, ImgFileAttachForm_State> {
    constructor(props) {
        super(props);
        this.state = {
            imageUrl: this.props.imageUrl,
        };
    }

    render() {
        const img = (
            <img
                className={
                    this.props.fileTitle == 'auth_header_logo'
                        ? 'sso-fields__image__header-logo'
                        : 'sso-fields__image__page-logo'
                }
                src={this.state.imageUrl}
                onError={() => {
                    this.setState({ imageUrl: null });
                }}
                alt="Sign in"
            />
        );

        return (
            <div className="sso-logo-pane-container">
                <div className="sso-fields__label">{this.props.text}</div>

                {this.props.canEdit ? (
                    <>
                        <div className="sso-fields__image-holder">
                            {this.state.imageUrl ? (
                                <div>
                                    <div>{img}</div>
                                    <div
                                        className={'sso-fields__delete-img-div'}
                                        onClick={() => {
                                            this.setState({ imageUrl: null });
                                            this.props.handleDeleteLogo(this.props.fileTitle);
                                        }}
                                    >
                                        <FontAwesomeIcon
                                            className={
                                                (this.props.fileTitle == 'auth_header_logo'
                                                    ? 'sso-fields__delete-img--header-logo'
                                                    : 'sso-fields__delete-img--page-logo') +
                                                ' sso-fields__delete-img clickable'
                                            }
                                            icon={faTimesCircle}
                                        />
                                    </div>
                                </div>
                            ) : (
                                <NoImageSelectedDisplay/>
                            )}
                        </div>

                        <div className="sso-fields__file-attachment" id={this.props.fileTitle}>
                            <FileAttachmentForm
                                index={this.props.index}
                                showLabel={false}
                                allowMultiple={false}
                                allowDirectories={false}
                                acceptedFormats=".jpg,.jpeg,.png,.gif,.tif"
                                showAcceptedFormats={false}
                                onFileChange={attachment => {
                                    this.props.onFileChange(attachment, this.props.fileTitle);
                                }}
                            />
                        </div>
                    </>
                ) : (
                    <div className="sso-fields__image-holder--view-only">
                        {this.state.imageUrl ? img : <NoImageSelectedDisplay/>}
                    </div>
                )}
            </div>
        );
    }
}

interface Props {
    headerLogoUrl: string;
    loginLogoUrl: string;
    onFileChange: Function;
    handleDeleteLogo: Function;
    canEdit: boolean;
}

export class SSOFields extends PureComponent<Props> {
    render() {
        return (
            <>
                <ImageAndFileAttachmentForm
                    text="Page Header Logo"
                    imageUrl={ActionURL.getBaseURL(true) + this.props.headerLogoUrl}
                    onFileChange={this.props.onFileChange}
                    handleDeleteLogo={this.props.handleDeleteLogo}
                    fileTitle="auth_header_logo"
                    canEdit={this.props.canEdit}
                    index={1}
                />

                <div className="sso-fields__spacer" />

                <ImageAndFileAttachmentForm
                    text="Login Page Logo"
                    imageUrl={ActionURL.getBaseURL(true) + this.props.loginLogoUrl}
                    onFileChange={this.props.onFileChange}
                    handleDeleteLogo={this.props.handleDeleteLogo}
                    fileTitle="auth_login_page_logo"
                    canEdit={this.props.canEdit}
                    index={2}
                />
            </>
        );
    }
}
