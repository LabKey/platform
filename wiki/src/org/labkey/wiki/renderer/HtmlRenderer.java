/*
 * Copyright (c) 2008-2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.labkey.wiki.renderer;

import org.apache.commons.lang3.StringUtils;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.labkey.api.attachments.Attachment;
import org.labkey.api.util.PageFlowUtil;
import org.labkey.api.util.TidyUtil;
import org.labkey.api.view.template.ClientDependency;
import org.labkey.api.wiki.WikiRenderer;
import org.labkey.api.wiki.FormattedHtml;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * User: Tamra Myers
 * Date: Aug 16, 2006
 * Time: 12:33:37 PM
 */
public class HtmlRenderer implements WikiRenderer
{
    private final String _hrefPrefix;
    private final String _attachPrefix;
    private final Map<String, String> _nameTitleMap;
    private final Map<String, Attachment> _attachments;

    private static Map<String, SubstitutionHandler> _substitutionHandlers = new HashMap<>();

    static
    {
        // Add SubstitutionHandlers for each ${labkey.<type>()}
        _substitutionHandlers.put("webPart", new WebPartSubstitutionHandler());
        _substitutionHandlers.put("dependency", new ClientDependencySubstitutionHandler());
    }


    public HtmlRenderer(String hrefPrefix, String attachPrefix, Map<String, String> nameTitleMap, @Nullable Collection<? extends Attachment> attachments)
    {
        _hrefPrefix = hrefPrefix;
        _attachPrefix = attachPrefix;
        _nameTitleMap = nameTitleMap == null ? new HashMap<>() : nameTitleMap;
        _attachments = new HashMap<>();

        if (null != attachments)
            for (Attachment a : attachments)
                _attachments.put(a.getName(), a);
    }
    

    public FormattedHtml format(String text)
    {
        LinkedList<String> errors = new LinkedList<>();
        if (text == null)
            return new FormattedHtml("");

        // Remove degenerate comments (e.g. "<!-->") as Tidy does not handle them properly
        text = text.replaceAll("<!--*>|<!>", "");

        FormattedHtml formattedHtml = handleLabKeySubstitutions(text);
        boolean volatilePage = formattedHtml.isVolatile();
        LinkedHashSet<ClientDependency> cds = new LinkedHashSet<>(formattedHtml.getClientDependencies());

        Document doc = TidyUtil.convertHtmlToDocument("<html><body>" + StringUtils.trimToEmpty(formattedHtml.getHtml()) + "</body></html>", false, errors);
        if (!errors.isEmpty() || doc == null)
        {
            StringBuilder innerHtml = new StringBuilder("<div class=\"labkey-error\"><b>An exception occurred while generating the HTML.  Please correct this content.</b></div><br>The error message was: ");
            if (!errors.isEmpty())
            {
                for (String error : errors)
                    innerHtml.append(PageFlowUtil.filter(error)).append("<hr>");
            }
            return new FormattedHtml(innerHtml.toString(), false, cds);
        }

        Set<String> wikiDependencies = new HashSet<>();

        // process A and IMG
        NodeList nl = doc.getElementsByTagName("a");
        Map<Element, String> linkExceptions = new HashMap<>();
        for (int i=0 ; i<nl.getLength() ; i++)
        {
            Element a = (Element)nl.item(i);
            try
            {
                String href = PageFlowUtil.decode(a.getAttribute("href"));

                Attachment at = _attachments.get(href);
                if (null != at)
                {
                    a.setAttribute("href", _attachPrefix + PageFlowUtil.encode(at.getName()));
                    wikiDependencies.add(at.getName());
                    continue;
                }

                if (href.isBlank() || href.startsWith("#"))
                    continue;

                String wikiName = StringUtils.substringBefore(href, "#");
                String title = _nameTitleMap.get(wikiName);
                if (null != title)
                {
                    // UNDONE: why is l.getName() null???
                    //a.setAttribute("href", _hrefPrefix + PageFlowUtil.encode(.getName()));
                    a.setAttribute("href", _hrefPrefix + PageFlowUtil.encode(href));
                    wikiDependencies.add(wikiName);
                    continue;
                }

                // if this doesn't look like a url, then link to it as a wiki page
                if (StringUtils.containsNone(href, "?:/.&"))
                {
                    a.setAttribute("href", _hrefPrefix + PageFlowUtil.encode(href));
                    wikiDependencies.add(wikiName);
                }
            }
            catch (IllegalArgumentException e)
            {
                // Issue 12160: there was an issue with decoding the href for the link, so store the exception text and link index for later
                linkExceptions.put(a, e.getMessage());
            }
        }

        nl = doc.getElementsByTagName("img");
        for (int i=0 ; i<nl.getLength() ; i++)
        {
            Element img = (Element)nl.item(i);
            String src = img.getAttribute("src");
            src = PageFlowUtil.decode(src);
            Attachment at = _attachments.get(src);
            if (null != at)
            {
                img.setAttribute("src", _attachPrefix + PageFlowUtil.encode(at.getName()));
                wikiDependencies.add(at.getName());
            }
        }

        // back to html
        StringBuilder innerHtml;

        try
        {
            innerHtml = new StringBuilder();

            //look for style elements, as tidy moves them to the head section
            NodeList styleNodes = doc.getElementsByTagName("style");
            for (int idx = 0; idx < styleNodes.getLength(); ++idx)
            {
                innerHtml.append(PageFlowUtil.convertNodeToHtml(styleNodes.item(idx)));
                innerHtml.append('\n');
            }

            Node bodyNode = doc.getElementsByTagName("body").item(0);
            // Uncomment the below to debug bodyNode contents
            // inspectNode(bodyNode, 0);

            String bodyHtml = PageFlowUtil.convertNodeToHtml(bodyNode);

            // Issue 12160: if there were any link decoding exceptions, replace the link with the error message
            if (linkExceptions.size() > 0)
            {
                for (Map.Entry<Element, String> link : linkExceptions.entrySet())
                {
                    String linkHtml = "<span class='labkey-error'>" + link.getValue() + "</span>";
                    bodyHtml = bodyHtml.replace(PageFlowUtil.convertNodeToHtml(link.getKey()), linkHtml);
                }
            }

            innerHtml.append(bodyHtml, "<body>".length(), bodyHtml.length()-"</body>".length());
        }
        catch (Exception e)
        {
            innerHtml = new StringBuilder("<div class=\"labkey-error\"><b>An exception occurred while generating the HTML.  Please correct this content.</b></div><br>The error message was: ");
            innerHtml.append(e.getMessage()); 
            volatilePage = false;
        }

        return new FormattedHtml(innerHtml.toString(), volatilePage, wikiDependencies, Collections.emptySet(), cds);
    }


    // ${labkey.<type>(<any_stream of characters>)}
    // Pattern.DOTALL allows the parameter list to span multiple lines
    private static final Pattern _substitutionPattern = Pattern.compile("\\$\\{labkey\\.(\\w+)\\((.*?)\\)\\}", Pattern.DOTALL);

    // <any word>='<any value>', allowing whitespace before, after, and in-between
    private static final Pattern _paramPattern = Pattern.compile("\\s*(\\w+)\\s*=\\s*'(.*)'\\s*");

    private FormattedHtml handleLabKeySubstitutions(String text)
    {
        if (text == null)
            return new FormattedHtml("");
        
        // Find all substitution templates embedded in wiki text that have the form ${labkey.<type>(<any_stream of characters>)}.
        Matcher webPartMatcher = _substitutionPattern.matcher(text);

        // If we find none, return immediately
        if (!webPartMatcher.find())
            return new FormattedHtml(text);

        List<Definition> definitions = new ArrayList<>(10);
        Map<Definition, List<String>> wikiErrors = new HashMap<>();
        do
        {
            List<String> paramErrors = new ArrayList<>();
            String substitutionType = webPartMatcher.group(1);          // type
            String params = webPartMatcher.group(2).replace(",", "");
            // Parse the parameters with the symbols in parseWith, they can be used in any order
            // as long as they are the same symbol starts and completes a parameter value
            List<String> paramList = new ArrayList<>();
            paramList.add(params);
            String[] parseWith = { "&#39;", "'" };
            for (String parser : parseWith)
            {
                List<String> paramListTemp = new ArrayList<>();
                for (String paramSection : paramList)
                {
                    String[] paramSplit = paramSection.split(parser);
                    for (int i = 0; i < paramSplit.length; i+=2)
                    {
                        paramListTemp.add(paramSplit[i] + (paramSplit.length > i + 1 ? "'" + paramSplit[i+1]  + "'": ""));
                    }
                }
                paramList = paramListTemp;
            }

            Map<String, String> paramMap = new HashMap<>(10);
            for (String param : paramList)
            {
                Matcher paramMatcher = _paramPattern.matcher(param);

                if (paramMatcher.matches())
                {
                    if (paramMap.containsKey(paramMatcher.group(1)))
                        paramErrors.add(param.trim() + ", there are multiple parameters with this name");
                    else
                        paramMap.put(paramMatcher.group(1), paramMatcher.group(2));
                }
                else if (param.trim().length() > 0)
                    paramErrors.add(param.trim());
            }

            // Stick new definition at beginning of list -- we want to replace them in reverse order
            Definition definition = new Definition(substitutionType, webPartMatcher.start(), webPartMatcher.end(), paramMap);
            definitions.add(0, definition);
            wikiErrors.put(definition, paramErrors);
        }
        while(webPartMatcher.find());

        StringBuilder sb = new StringBuilder(text);
        boolean volatilePage = false;
        LinkedHashSet<ClientDependency> cds = new LinkedHashSet<>();

        // Get the corresponding substitution handler for each type and replace template with substitution
        for (Definition definition : definitions)
        {
            SubstitutionHandler handler = _substitutionHandlers.get(definition.getType());
            FormattedHtml substitution;

            if (null != handler)
                substitution = handler.getSubstitution(definition.getParams());
            else
                substitution = new FormattedHtml("<br><font class='error' color='red'>Error: unknown type, \"labkey." + definition.getType() + "\"</font>");

            sb.replace(definition.getStart(), definition.getEnd(), substitution.getHtml());

            if (substitution.isVolatile())
                volatilePage = true;

            cds.addAll(substitution.getClientDependencies());

            List<String> paramErrors = wikiErrors.get(definition);
            if (paramErrors.size() > 0)
            {
                String errorHTML = "<br>";
                for (String error : paramErrors)
                    errorHTML = errorHTML.concat("<font class='error' color='red'>Error with parameter " +
                            error + " in " + definition.getType() + "</font><br><br>");
                sb.insert(definition.getStart() + substitution.getHtml().length(), errorHTML);
            }
        }

        return new FormattedHtml(sb.toString(), volatilePage, cds);
    }


    private static class Definition
    {
        private final String _type;
        private final int _start;
        private final int _end;
        private final Map<String, String> _params;

        private Definition(String type, int start, int end, Map<String, String> params)
        {
            _type = type;
            _start = start;
            _end = end;
            _params = params;
        }

        public String getType()
        {
            return _type;
        }

        private int getEnd()
        {
            return _end;
        }

        private Map<String, String> getParams()
        {
            return _params;
        }

        private int getStart()
        {
            return _start;
        }
    }


    interface SubstitutionHandler
    {
        @NotNull FormattedHtml getSubstitution(Map<String, String> params);
    }
}
