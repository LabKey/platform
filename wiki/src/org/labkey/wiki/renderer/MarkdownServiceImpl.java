/*
 * Copyright (c) 2017-2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.labkey.wiki.renderer;

import org.labkey.api.markdown.MarkdownService;
import org.labkey.api.module.Module;
import org.labkey.api.module.ModuleLoader;
import org.labkey.api.reports.LabkeyScriptEngineManager;
import org.labkey.api.resource.Resource;
import org.labkey.api.services.ServiceRegistry;
import org.labkey.api.util.Path;
import org.labkey.api.util.StringUtilsLabKey;

import javax.script.Invocable;
import javax.script.ScriptEngine;
import javax.script.ScriptException;
import java.io.BufferedInputStream;
import java.io.InputStreamReader;

public class MarkdownServiceImpl implements MarkdownService
{
    private Object mdCompiled;
    private Invocable invocable;

    public MarkdownServiceImpl() throws Exception
    {
        ScriptEngine engine;
        LabkeyScriptEngineManager svc = ServiceRegistry.get().getService(LabkeyScriptEngineManager.class);
        engine = svc.getEngineByName("nashorn");

        Module module = ModuleLoader.getInstance().getModule("Wiki");
        Path path = Path.parse("scripts/").append("markdown-it.js");
        Resource r = module.getModuleResource(path);
        if (r != null && r.isFile() && r.exists())
        {
            engine.eval(new InputStreamReader(new BufferedInputStream(r.getInputStream(), 1024 * 20), StringUtilsLabKey.DEFAULT_CHARSET));
            engine.eval("var md = new markdownit({" +
                    "breaks: true," +
                    "linkify: true" +
                    "})");
            mdCompiled = engine.eval("md");
            invocable = (Invocable) engine;
            invocable.invokeMethod(mdCompiled, "render", "# call render method here to ensure that nashorn compiles this method before use by app");
        }

    }

    @Override
    public String toHtml(String mdText) throws NoSuchMethodException, ScriptException
    {
        // make sure that the source text has the carriage returns escaped and the whole thing encoded
        // otherwise it wont parse right as a js string if it hits a cr or a quote.
        //        mdText = mdText.replace("\n", "\\n");
        //        mdText = PageFlowUtil.filter(mdText);
        //        Object testResult = engine.eval("md.render('" + mdText + "')");

        // Better yet, avoid js injection by using the invokeMethod syntax that takes a java String as a param

        // markdownit wont accept null as input param
        if (null == mdText)
            mdText = "";
        Object html = invocable.invokeMethod(mdCompiled, "render", mdText);
        if (null == html)
            return null;
        // #32468 include selector so we can have markdown-specific styling namespace
        return "<div class=\"lk-markdown-container\">" + html.toString() + "</div>";
    }
}
