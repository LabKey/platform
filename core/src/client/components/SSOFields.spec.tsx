import React from 'react';
import {SSOFields, ImageAndFileAttachmentForm} from './SSOFields';
import renderer from 'react-test-renderer';
import {shallow} from "enzyme";

const testImageUrl = "https://en.wikipedia.org/wiki/Moose#/media/File:Moose_superior.jpg";

describe("<SSOFields/>", () => {
    test("No images attached", () => {
        const component =
            <SSOFields
                canEdit={true}
                headerLogoUrl={null}
                loginLogoUrl={null}
                onFileChange={() => {}}
                handleDeleteLogo={() => {}}
            />;

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("Editable mode", () => {
        const component =
            <SSOFields
                canEdit={true}
                headerLogoUrl={null}
                loginLogoUrl={null}
                onFileChange={() => {}}
                handleDeleteLogo={() => {}}
            />;

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("View-only mode", () => {
        const component =
            <SSOFields
                canEdit={false}
                headerLogoUrl={null}
                loginLogoUrl={null}
                onFileChange={() => {}}
                handleDeleteLogo={() => {}}
            />;

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("<ImageAndFileAttachmentForm/>", () => {
    test("Click remove-image button", () => {
        const component =
            <ImageAndFileAttachmentForm
                text={"Page Header Logo"}
                imageUrl={testImageUrl}
                onFileChange={() => {}}
                handleDeleteLogo={() => {}}
                fileTitle={"auth_header_logo"}
                canEdit={true}
                index={1}
            />;
        const wrapper = shallow<SSOFields>(component);

        const deleteImgButton = wrapper.find('.sso-fields__delete-img-div');
        deleteImgButton.simulate('click');

        expect(wrapper.state()).toHaveProperty('imageUrl', null)
    });
});
