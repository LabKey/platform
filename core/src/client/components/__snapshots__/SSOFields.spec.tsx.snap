// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SSOFields/> Editable mode 1`] = `
Array [
  <div
    className="sso-logo-pane-container"
  >
    <div
      className="sso-fields__label"
    >
      Page Header Logo
    </div>
    <div
      className="sso-fields__image-holder"
    >
      <div>
        <div>
          <img
            alt="Sign in"
            className="sso-fields__image__header-logo"
            onError={[Function]}
            src="http://localhostnull"
          />
        </div>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times-circle fa-w-16 sso-fields__delete-img--header-logo sso-fields__delete-img clickable"
          color="#d9534f"
          data-icon="times-circle"
          data-prefix="fas"
          focusable="false"
          onClick={[Function]}
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <div
      className="sso-fields__file-attachment"
      id="auth_header_logo"
    >
      <span
        className="translator--toggle__wizard"
      >
        <div
          className="row"
        >
          <div
            className="col-sm-12"
          >
            <div
              className="wizard-row--container"
            >
              <div>
                <div
                  className="file-upload--container block"
                >
                  <label
                    className="file-upload--label"
                    htmlFor="fileUpload1"
                    onDragEnter={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDrop={[Function]}
                  >
                    <i
                      aria-hidden="true"
                      className="fa fa-cloud-upload fa-2x cloud-logo"
                    />
                    Select file or drag and drop here.
                  </label>
                  <input
                    accept=".jpg,.jpeg,.png,.gif,.tif"
                    className="file-upload--input"
                    id="fileUpload1"
                    multiple={false}
                    name="fileUpload1"
                    onChange={[Function]}
                    type="file"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>,
  <div
    className="sso-fields__spacer"
  />,
  <div
    className="sso-logo-pane-container"
  >
    <div
      className="sso-fields__label"
    >
      Login Page Logo
    </div>
    <div
      className="sso-fields__image-holder"
    >
      <div>
        <div>
          <img
            alt="Sign in"
            className="sso-fields__image__page-logo"
            onError={[Function]}
            src="http://localhostnull"
          />
        </div>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times-circle fa-w-16 sso-fields__delete-img--page-logo sso-fields__delete-img clickable"
          color="#d9534f"
          data-icon="times-circle"
          data-prefix="fas"
          focusable="false"
          onClick={[Function]}
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <div
      className="sso-fields__file-attachment"
      id="auth_login_page_logo"
    >
      <span
        className="translator--toggle__wizard"
      >
        <div
          className="row"
        >
          <div
            className="col-sm-12"
          >
            <div
              className="wizard-row--container"
            >
              <div>
                <div
                  className="file-upload--container block"
                >
                  <label
                    className="file-upload--label"
                    htmlFor="fileUpload2"
                    onDragEnter={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDrop={[Function]}
                  >
                    <i
                      aria-hidden="true"
                      className="fa fa-cloud-upload fa-2x cloud-logo"
                    />
                    Select file or drag and drop here.
                  </label>
                  <input
                    accept=".jpg,.jpeg,.png,.gif,.tif"
                    className="file-upload--input"
                    id="fileUpload2"
                    multiple={false}
                    name="fileUpload2"
                    onChange={[Function]}
                    type="file"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>,
]
`;

exports[`<SSOFields/> No images attached 1`] = `
Array [
  <div
    className="sso-logo-pane-container"
  >
    <div
      className="sso-fields__label"
    >
      Page Header Logo
    </div>
    <div
      className="sso-fields__image-holder"
    >
      <div>
        <div>
          <img
            alt="Sign in"
            className="sso-fields__image__header-logo"
            onError={[Function]}
            src="http://localhostnull"
          />
        </div>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times-circle fa-w-16 sso-fields__delete-img--header-logo sso-fields__delete-img clickable"
          color="#d9534f"
          data-icon="times-circle"
          data-prefix="fas"
          focusable="false"
          onClick={[Function]}
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <div
      className="sso-fields__file-attachment"
      id="auth_header_logo"
    >
      <span
        className="translator--toggle__wizard"
      >
        <div
          className="row"
        >
          <div
            className="col-sm-12"
          >
            <div
              className="wizard-row--container"
            >
              <div>
                <div
                  className="file-upload--container block"
                >
                  <label
                    className="file-upload--label"
                    htmlFor="fileUpload1"
                    onDragEnter={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDrop={[Function]}
                  >
                    <i
                      aria-hidden="true"
                      className="fa fa-cloud-upload fa-2x cloud-logo"
                    />
                    Select file or drag and drop here.
                  </label>
                  <input
                    accept=".jpg,.jpeg,.png,.gif,.tif"
                    className="file-upload--input"
                    id="fileUpload1"
                    multiple={false}
                    name="fileUpload1"
                    onChange={[Function]}
                    type="file"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>,
  <div
    className="sso-fields__spacer"
  />,
  <div
    className="sso-logo-pane-container"
  >
    <div
      className="sso-fields__label"
    >
      Login Page Logo
    </div>
    <div
      className="sso-fields__image-holder"
    >
      <div>
        <div>
          <img
            alt="Sign in"
            className="sso-fields__image__page-logo"
            onError={[Function]}
            src="http://localhostnull"
          />
        </div>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times-circle fa-w-16 sso-fields__delete-img--page-logo sso-fields__delete-img clickable"
          color="#d9534f"
          data-icon="times-circle"
          data-prefix="fas"
          focusable="false"
          onClick={[Function]}
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <div
      className="sso-fields__file-attachment"
      id="auth_login_page_logo"
    >
      <span
        className="translator--toggle__wizard"
      >
        <div
          className="row"
        >
          <div
            className="col-sm-12"
          >
            <div
              className="wizard-row--container"
            >
              <div>
                <div
                  className="file-upload--container block"
                >
                  <label
                    className="file-upload--label"
                    htmlFor="fileUpload2"
                    onDragEnter={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDrop={[Function]}
                  >
                    <i
                      aria-hidden="true"
                      className="fa fa-cloud-upload fa-2x cloud-logo"
                    />
                    Select file or drag and drop here.
                  </label>
                  <input
                    accept=".jpg,.jpeg,.png,.gif,.tif"
                    className="file-upload--input"
                    id="fileUpload2"
                    multiple={false}
                    name="fileUpload2"
                    onChange={[Function]}
                    type="file"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>,
]
`;

exports[`<SSOFields/> View-only mode 1`] = `
Array [
  <div
    className="sso-logo-pane-container"
  >
    <div
      className="sso-fields__label"
    >
      Page Header Logo
    </div>
    <div
      className="sso-fields__image-holder--view-only"
    >
      <img
        alt="Sign in"
        className="sso-fields__image__header-logo"
        onError={[Function]}
        src="http://localhostnull"
      />
    </div>
  </div>,
  <div
    className="sso-fields__spacer"
  />,
  <div
    className="sso-logo-pane-container"
  >
    <div
      className="sso-fields__label"
    >
      Login Page Logo
    </div>
    <div
      className="sso-fields__image-holder--view-only"
    >
      <img
        alt="Sign in"
        className="sso-fields__image__page-logo"
        onError={[Function]}
        src="http://localhostnull"
      />
    </div>
  </div>,
]
`;
